src/streamlink_cli/utils/progress.py

```diff
@@ -1,153 +1,161 @@
+
+    @classmethod
+    def format_time(cls, elapsed: float) -> str:
+        if elapsed < 0:
+            elapsed = 0
+
+        hours = ""
+        minutes = ""
+
+        if elapsed >= 3600:
+            hours = f"{int(elapsed / (60 * 60))}h"
+        if elapsed >= 60:
+            if elapsed >= 3600:
+                minutes = f"{int((elapsed % (60 * 60)) / 60):02d}m"
             else:
-                return "{0:3.1f} {1}".format(size, suffix)
-
-        size /= 1024.0
-
-
-def format_time(elapsed):
-    """Formats elapsed seconds into a human readable format."""
-    hours = int(elapsed / (60 * 60))
-    minutes = int((elapsed % (60 * 60)) / 60)
-    seconds = int(elapsed % 60)
-
-    rval = ""
-    if hours:
-        rval += "{0}h".format(hours)
-
-    if elapsed > 60:
-        rval += "{0}m".format(minutes)
-
-    rval += "{0}s".format(seconds)
-    return rval
-
-
-def create_status_line(**params):
-    """Creates a status line with appropriate size."""
-    max_size = get_terminal_size().columns - 1
-
-    for fmt in PROGRESS_FORMATS:
-        status = fmt.format(**params)
-
-        if len(status) <= max_size:
-            break
-
-    return status
-
-
-def progress(iterator, prefix):
-    """Progress an iterator and updates a pretty status line to the terminal.
-
-    The status line contains:
-     - Amount of data read from the iterator
-     - Time elapsed
-     - Average speed, based on the last few seconds.
-    """
-    if terminal_width(prefix) > 25:
-        prefix = (".." + get_cut_prefix(prefix, 23))
-    speed_updated = start = time()
-    speed_written = written = 0
-    speed_history = deque(maxlen=5)
-
-    for data in iterator:
-        yield data
-
-        now = time()
-        elapsed = now - start
-        written += len(data)
-
-        speed_elapsed = now - speed_updated
-        if speed_elapsed >= 0.5:
-            speed_history.appendleft((
-                written - speed_written,
-                speed_updated,
-            ))
-            speed_updated = now
-            speed_written = written
-
-            speed_history_written = sum(h[0] for h in speed_history)
-            speed_history_elapsed = now - speed_history[-1][1]
-            speed = speed_history_written / speed_history_elapsed
-
-            status = create_status_line(
-                prefix=prefix,
-                written=format_filesize(written),
-                elapsed=format_time(elapsed),
-                speed=format_filesize(speed)
+                minutes = f"{int((elapsed % (60 * 60)) / 60):1d}m"
+
+        if elapsed >= 60:
+            return f"{hours}{minutes}{int(elapsed % 60):02d}s"
+        else:
+            return f"{hours}{minutes}{int(elapsed % 60):1d}s"
+
+
+class Progress(Thread):
+    def __init__(
+        self,
+        output: Optional[TerminalOutput] = None,
+        formatter: Optional[ProgressFormatter] = None,
+        interval: float = 0.25,
+        history: int = 20,
+        threshold: int = 2,
```

Comment:
These values could be tweaked.

Four updates per seconds seems reasonable, and so is keeping a download speed history (the result is the average in this time frame) of twenty seconds which ensures that segmented streams with large segment durations (but still less than 20 seconds) don't make the download speed go up and down in between segment downloads and segment queue wait times.

Reply:
Let's see if anyone has feedback after the next release. If so they can be modified but these seem reasonable to me.